tasks/puppet_infra_upgrade.rb

```diff
@@ -25,4 +31,50 @@ def main
   end
 end
 
+def inventory_uri
+  @inventory_uri ||= URI.parse('https://localhost:8143/orchestrator/v1/inventory')
+end
+
+def request_object(nodes:, token_file:)
+  token = File.read('/root/.puppetlabs/token')
+  body = {
+    'nodes' => nodes,
+  }.to_json
+
+  request = Net::HTTP::Post.new(inventory_uri.request_uri)
+  request['Content-Type'] = 'application/json'
+  request['X-Authentication'] = token
+  request.body = body
+
+  request
+end
+
+def http_object
+  http = Net::HTTP.new(inventory_uri.host, inventory_uri.port)
+  http.use_ssl = true
+  http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+
+  http
+end
+
+def wait_until_connected(nodes:, token_file:, timeout: 120)
+  http = http_object
+  request = request_object(nodes: nodes, token_file: token_file)
+  inventory = {}
+  Timeout::timeout(timeout) do
```

Comment:
In this article: https://medium.com/@adamhooper/in-ruby-dont-use-timeout-77d9d4e5a001 the dangers of Timeout::timeout are outlined. Is the worry that this code might raise for unexpected conditions justified?

Reply:
In this code, we only care about the assignment operation to the variable `inventory`. The objective isn't to hard-lock execution time to a particular interval, it's just "stop trying after _n_ seconds if you haven't succeeded". As such, I believe the behavior of Timeout here will not cause anything unexpected to happen.